src/celeritas/neutron/model/detail/NuclearZoneBuilder.hh

```diff
@@ -0,0 +1,410 @@
+                         target.neutron_loss_energy().value()};
+
+    for (auto i : range(num_of_zones))
+    {
+        for (auto ptype : range(dim))
+        {
+            components[i].density[ptype]
+                = static_cast<real_type>(num_nucleons[ptype]) * integral[i]
+                  / (total_integral * components[i].volume);
+            components[i].fermi_mom[ptype]
+                = options_.fermi_scale
+                  * std::cbrt(components[i].density[ptype]);
+            components[i].potential[ptype]
+                = this->half() * ipow<2>(components[i].fermi_mom[ptype])
+                      / mass[ptype]
+                  + dm[ptype];
+        }
+    }
+
+    integral.clear();
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Calculate the nuclear radius (R) computed from the atomic mass number (A).
+ *
+ * For \f$ A > 4 \f$, the nuclear radius with two parameters takes the form,
+ * \f[
+     R = [ 1.16 * A^{1/3} - 1.3456 / A^{1/3} ] \cdot R_{scale}
+   \f]
+ * where \f$ R_{scale} \f$ is a configurable parameter in [femtometer], while
+ * \f$ R = 1.2 A^{1/3} \cdot R_{scale} \f$ (default) with a single parameter.
+ */
+real_type NuclearZoneBuilder::calc_nuclear_radius(AtomicMassNumber a)
+{
+    // Nuclear radius computed from A
+    real_type nuclear_radius{0};
+    auto amass = a.get();
+
+    if (amass > 4)
+    {
+        real_type cbrt_a = std::cbrt(static_cast<real_type>(amass));
+        real_type par_a = (options_.use_two_params ? 1.16 : 1.2);
+        real_type par_b = (options_.use_two_params ? -1.3456 : 0);
+        nuclear_radius = options_.radius_scale
+                         * (par_a * cbrt_a + par_b / cbrt_a);
+    }
+    else
+    {
+        nuclear_radius = options_.radius_small
+                         * (amass == 4 ? options_.radius_alpha : 1);
+    }
+
+    return nuclear_radius;
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Integrate the Woods-Saxon potential, \f$ V(r) \f$,
+ *
+ * \f[
+     V(r) = frac{V_{o}}{1 + e^{\frac{r - R}{a}}}
+   \f]
+ * numerically over the volume from \f$ r_{min} \f$ to \f$ r_{rmax} \f$, where
+ * \f$ V_{o}, R, a\f$ are the potential well depth, nuclear radius, and
+ * surface thickness (skin depth), respectively.
+ */
+real_type
+NuclearZoneBuilder::integrate_woods_saxon(real_type rmin,
+                                          real_type rmax,
+                                          real_type nuclear_radius) const
+{
+    real_type skin_ratio = nuclear_radius / skin_depth_;
+
+    real_type ws_shift = 2 * skin_ratio;
+    real_type delta_r = rmax - rmin;
+
+    // Average value of \f$ r^{2} V(r) dr \f$ at boundaries
+    real_type integral = half() * delta_r
+                         * (rmin * (rmin + ws_shift) / (1 + std::exp(rmin))
+                            + rmax * (rmax + ws_shift) / (1 + std::exp(rmax)));
```

Comment:
This (and the gaussian case below) are just initial guesses for the integral... this should be trivial to incorporate into the adaptive quadrature solver itself, right?

Reply:
You are absolutely right. This was a lazy man's implementation, but definitely can be integrated into `PotentialIntegrator` that you suggested early.